client/src/components/ResponseHandler.tsx

```diff
@@ -24,25 +38,54 @@ function ResponseHandler() {
     fetchData();
   }, []);
 
-  // Send `userInput` to the server, receive response, then add to `conversation`.
-  const handleUserInput = async (userInput: string) => {
+  useEffect(() => {
+    if (gameID) {
+      handleUserInput("");
+    }
+  }, [gameID]);
+
+
+  const sendRequest = async (userInput: string) => {
     const response = await fetch("http://127.0.0.1:5000/play/" + gameID, {
       method: "POST",
       headers: { "Content-Type": "application/json" },
       body: JSON.stringify({ input: userInput }),
     });
-
     if (response.ok) {
       const data = await response.json();
-      let message = data.response?.message ?? "";
-      let options = (data.response?.options ?? []).map((option: any[]) => option.join(': '));
-
       console.log(data);
-      setConversation((prev) => [...prev, "\nUser: "+userInput+"\n", message, ...options]);
-      //TODO: add expects-response filter
-    } else {
-      console.error("Failed to send the message");
+      return data;
     }
+    return Error("Error: Server Request Failed.");
+  };
+
+  const handleUserInput = async (userInput: string) => {
+    if (lastMessage?.type == "SelectOptionCommand" && userInput !== "") {
+      let last: SelectOptionCommand = lastMessage as SelectOptionCommand;
+      if (!last.options.some((tuple) => tuple.includes(userInput))) {
+        let error = "Invalid option. Please try again.";
+        setConversation((prev) => [...prev, error]);
+        return;
+      }
+    }
+
+    let data = await sendRequest(userInput);
+    let response = castCommand(data.response);
+    let styles: TextStyles = extractTextStyles(response);
+    let text: string[] = extractTextContent(response);
+
+    console.log("resp", response);
+
+    let uText = userInput ? "\nYou: " + userInput : "";
+    setConversation((prev) => [...prev, uText, ...text]);
+
+    setStyles(styles);
+    setLastMessage(
+      response.type == "SelectOptionCommand"
+        ? (response as SelectOptionCommand)
+        : undefined
+    );
```

Comment:
So we're only doing this if it was a SelectOptionCommand?

Reply:
We only need to set it if its an option command, otherwise we set to undefined. We could also set it to be else: response as BaseCommand, but we dont use lastMessage for anything else

Reply:
Sounds good.